File: generate_document/generate_document.ts
function generateDocument(characters: string, document: string): boolean {
    const charCount: Record<string, number> = {};

    for (const c of characters) {
        charCount[c] = (charCount[c] || 0) + 1;
    }

    for (const c of document) {
        if (!charCount[c] || charCount[c] === 0) {
            return false;
        }
        charCount[c] -= 1;
    }

    return true;
}

console.log(generateDocument("abc", "cab"));

File: generate_document/generate_document.go
package main

import (
    "fmt"
)

func generateDocument(characters, document string) bool {
  charCount := make(map[rune]int)

  for _, c := range characters {
    charCount[c]++
  }

  for _, c := range document {
    if count, ok := charCount[c]; !ok || count == 0 {
        return false
    }
    charCount[c]--
  }

  return true
}

func main() {
    fmt.Println(generateDocument("abc", "cab"))
}
